pkg/sfu/bwe/remotebwe/remote_bwe.go

```diff
@@ -0,0 +1,391 @@
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	return r.bweListener
+}
+
+func (r *RemoteBWE) Reset() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.channelObserver = r.newChannelObserverNonProbe()
+}
+
+func (r *RemoteBWE) Stop() {
+	r.stop.Break()
+}
+
+func (r *RemoteBWE) HandleREMB(
+	receivedEstimate int64,
+	isProbeFinalizing bool,
+	expectedBandwidthUsage int64,
+	sentPackets uint32,
+	repeatedNacks uint32,
+) {
+	r.lock.Lock()
+	r.lastReceivedEstimate = receivedEstimate
+	r.lastExpectedBandwidthUsage = expectedBandwidthUsage
+
+	if !isProbeFinalizing {
+		r.channelObserver.AddEstimate(r.lastReceivedEstimate)
+		r.channelObserver.AddNack(sentPackets, repeatedNacks)
+	}
+
+	var (
+		shouldNotify             bool
+		state                    bwe.CongestionState
+		committedChannelCapacity int64
+	)
+	if !r.isInProbe {
+		shouldNotify, state, committedChannelCapacity = r.congestionDetectionStateMachine()
+	}
+	r.lock.Unlock()
+
+	if shouldNotify {
+		if bweListener := r.getBWEListener(); bweListener != nil {
+			bweListener.OnCongestionStateChange(state, committedChannelCapacity)
+		}
+	}
+}
+
+func (r *RemoteBWE) congestionDetectionStateMachine() (bool, bwe.CongestionState, int64) {
+	newState := r.congestionState
+	update := false
+	trend, reason := r.channelObserver.GetTrend()
+	switch r.congestionState {
+	case bwe.CongestionStateNone:
+		if trend == bwe.ChannelTrendCongesting {
+			if r.estimateAvailableChannelCapacity(reason) {
+				newState = bwe.CongestionStateCongested
+			}
+		}
+
+	case bwe.CongestionStateCongested:
+		if trend == bwe.ChannelTrendCongesting {
+			if r.estimateAvailableChannelCapacity(reason) {
+				// update state sa this needs to reset switch time to wait for congestion min duration again
+				update = true
+			}
+		} else if time.Since(r.congestionStateSwitchedAt) >= r.params.Config.CongestedMinDuration {
+			newState = bwe.CongestionStateNone
+		}
+	}
+
+	shouldNotify := false
+	if newState != r.congestionState || update {
+		r.updateCongestionState(newState, reason)
+		shouldNotify = true
+	}
+
+	return shouldNotify, r.congestionState, r.committedChannelCapacity
+}
+
+func (r *RemoteBWE) estimateAvailableChannelCapacity(reason channelCongestionReason) bool {
+	var estimateToCommit int64
+	switch reason {
+	case channelCongestionReasonLoss:
+		estimateToCommit = int64(float64(r.lastExpectedBandwidthUsage) * (1.0 - r.params.Config.NackRatioAttenuator*r.channelObserver.GetNackRatio()))
+	default:
+		estimateToCommit = r.lastReceivedEstimate
+	}
+	if estimateToCommit > r.lastReceivedEstimate {
+		estimateToCommit = r.lastReceivedEstimate
+	}
+
+	commitThreshold := int64(r.params.Config.ExpectedUsageThreshold * float64(r.lastExpectedBandwidthUsage))
+	action := "applying"
+	if estimateToCommit > commitThreshold {
+		action = "skipping"
+	}
+
+	if action == "applying" {
+		r.params.Logger.Infow(
+			fmt.Sprintf("remote bwe: channel congestion detected, %s channel capacity update", action),
```

Comment:
these sprintfs are only ever called with 1 unique value each?

Reply:
Will make an unlikely logger. Yeah, the difference is mainly `Infow` vs `Debugw`.